components/builder-scheduler/src/server/scheduler.rs

```diff
@@ -420,39 +381,53 @@ impl ScheduleMgr {
     }
 
     fn process_status(&mut self) -> Result<()> {
-        self.status_sock.recv(&mut self.msg, 0)?;
-        let job: Job = parse_from_bytes(&self.msg)?;
-        let group: proto::Group = self.get_group(job.get_owner_id())?;
-
-        self.logger.log_group_job(&group, &job);
-
-        match self.datastore.set_group_job_state(&job) {
-            Ok(_) => {
-                if job.get_state() == jobsrv::JobState::Failed {
-                    match self.skip_projects(&group, job.get_project().get_name()) {
-                        Ok(_) => (),
-                        Err(e) => {
-                            warn!(
-                                "Error skipping projects for {:?}: {:?}",
-                                job.get_project().get_name(),
-                                e
-                            );
-                        }
-                    };
-                }
+        loop {
+            // Take the top job status from queue
+            let mut jobs = self.datastore.get_job_status(1)?;
 
-                match job.get_state() {
-                    jobsrv::JobState::Complete => {
-                        self.promote_to_sandbox(&job)?;
-                        self.update_group_state(job.get_owner_id())?
+            if jobs.len() == 0 {
+                break;
+            }
+
+            // This unwrap is fine, because we just checked our length
+            let job = jobs.pop().unwrap();
+            info!("Got job status {:?}", job);
+
+            let group: proto::Group = self.get_group(job.get_owner_id())?;
+
+            self.logger.log_group_job(&group, &job);
+
+            match self.datastore.set_group_job_state(&job) {
+                Ok(_) => {
+                    if job.get_state() == jobsrv::JobState::Failed {
+                        match self.skip_projects(&group, job.get_project().get_name()) {
+                            Ok(_) => (),
+                            Err(e) => {
+                                warn!(
+                                    "Error skipping projects for {:?}: {:?}",
+                                    job.get_project().get_name(),
+                                    e
+                                );
+                            }
+                        };
                     }
 
-                    jobsrv::JobState::Failed => self.update_group_state(job.get_owner_id())?,
+                    match job.get_state() {
+                        jobsrv::JobState::Complete => {
+                            self.promote_to_sandbox(&job)?;
+                            self.update_group_state(job.get_owner_id())?
+                        }
 
-                    _ => (),
+                        jobsrv::JobState::Failed => self.update_group_state(job.get_owner_id())?,
+
+                        _ => (),
+                    }
+
+                    // Clear the top job status from the queue
+                    self.datastore.clear_job_status(1)?;
```

Comment:
This seems like it could have the potential for subtle bugs; can we just remove, by ID, the entry for the job we just processed, instead of assuming that it's at the top of our queue?